### PR TITLE
35 group

### PR DESCRIPTION
**Pull request by derekjn**
Consumers can now be started with a `group_id`, which causes offsets to be stored in brokers.
- [x] Force offset commit in `save_consumer_offsets`
- [x] Respect `offset` parameter

### CODE REVIEW

File: pipeline_kafka.c [hunk 1]
```diff
@@ -1176,8 +1214,55 @@ kafka_consume_main(Datum arg)
 	if (broker_version)
 		rd_kafka_conf_set(conf, "broker.version.fallback", broker_version, NULL, 0);
 
-	kafka = rd_kafka_new(RD_KAFKA_CONSUMER, conf, err_msg, sizeof(err_msg));
-	rd_kafka_set_logger(kafka, consumer_logger);
+	if (consumer.group_id)
+	{
+		rd_kafka_conf_set(conf, "group.id", consumer.group_id, NULL, 0);
+		rd_kafka_conf_set(conf, "offset.store.method", "broker", NULL, 0);
+		rd_kafka_topic_conf_set(topic_conf, "topic.auto.offset.reset", "latest", NULL, 0);
```
Comment (usmanm) [inline pipeline_kafka.c:1280]: This should be whatever the user specifies as start offset? Maybe they want it to be oldest.

File: pipeline_kafka.c [hunk 2]
```diff
@@ -1176,8 +1214,55 @@ kafka_consume_main(Datum arg)
...
+	}
+
+	/*
+	 * Override consumer configuration with anying specified by pipeline_kafka.consumer_config
+	 */
+	if (consumer_config)
```
Comment (usmanm) [inline pipeline_kafka.c:1286]: Factor out into a static function. This bad is becoming too long.

**Comment by usmanm** [comment]
Where are we committing the offsets into Kafka? They're committed periodically but we should commit them manually at the end of each loop.

**Review by usmanm — APPROVED** [review]

File: pipeline_kafka.c (hunk 1 above)
Comment (derekjn) [reply pipeline_kafka.c:1280]: This is just the default. Users can override any of this with `pipeline_kafka.consumer_config = 'topic.auto.offset.reset=whatever,...`

Comment (usmanm) [reply pipeline_kafka.c:1280]: No, but we take a `start_offset` parameter in the `begin_consume` function. I think that should be respected in this mode as well.

Comment (derekjn) [reply pipeline_kafka.c:1280]: This parameter is only used when there are no existing offsets stored in the broker for the given group. I see what you're saying though, which is that we should respect the offset parameter even when reading from broker-based offsets. I'll add that.

**Comment by derekjn** [comment]
> Where are we committing the offsets into Kafka? They're committed periodically but we should commit them manually at the end of each loop.

`librdkafka` does it based on `auto.commit.interval.ms` (and on close). I'll force a commit in `save_consumer_offsets` though as well.

**Comment by usmanm** [comment]
We should increase the `auto.commit.interval.ms` to be really high so it never auto-commits. We should only commit after the batch has been processed.

**Comment by derekjn** [comment]
> We should increase the auto.commit.interval.ms to be really high so it never auto-commits. We should only commit after the batch has been processed.

I'll just disable it. `auto.commit.enable=false`.

**Pull request closed without merging**